app/controllers/concerns/shopify_app/domain_protection.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+module ShopifyApp
+  module DomainProtection
+    extend ActiveSupport::Concern
+
+    included do
+      before_action :check_shop_domain
+      before_action :check_shop_known
+    end
+
+    def current_shopify_domain
+      return if params[:shop].blank?
+      @shopify_domain ||= ShopifyApp::Utils.sanitize_shop_domain(params[:shop])
+    end
+
+    private
+
+    def check_shop_domain
+      redirect_to(ShopifyApp.configuration.login_url) unless current_shopify_domain
+    end
+
+    def check_shop_known
+      @shop = SessionRepository.retrieve_shop_session_by_shopify_domain(current_shopify_domain)
```

Comment:
This looks good now!

`retrieve_shop_session_by_domain` ~> `retrieve_shop_session_by_shopify_domain`